Extra process guards

heckj opened the pull request:
Following on from earlier PR (https://github.com/apple/swift-package-manager/pull/921) and bug SR-3275 - adding in the mechanisms to support the extra guard logic for validating that an executable file (i.e. git, etc) exists, and making the results when it doesn't exist more consistent across all the various underlying toolchains.

Here for visibility and feedback...

Next steps:
- [x] rebase after https://github.com/apple/swift-package-manager/pull/1002 & https://github.com/apple/swift-package-manager/pull/1003 merged
- [x] add in string output handling for new errors enums in Commands/Errors.swift
- [x] add in relevant tests to validate

aciidb0mb3r commented on Sources/Basic/FileSystem.swift:
```diff
@@ -101,6 +101,9 @@ public protocol FileSystem {
     func isFile(_ path: AbsolutePath) -> Bool
+    
+    /// Check to determine whether the given path is an accessible and executable file
+    func isExecutableFile(_ path: AbsolutePath) -> Bool
```
Please create a separate PR for this change. We try to keep all orthogonal changes (especially API addition in Basic) in different PRs.

heckj replied:
will do...

heckj replied:
migrated the Filesystem protocol extension, implementation and related tests to another PR: https://github.com/apple/swift-package-manager/pull/997

aciidb0mb3r commented on Sources/Basic/FileSystem.swift:
```diff
@@ -155,6 +158,18 @@ public extension FileSystem {
 /// Concrete FileSystem implementation which communicates with the local file system.
 private class LocalFileSystem: FileSystem {
+    
+    /// Check to determine whether the given path is an accessible and executable file
```
Don't require the doc comment here.

heckj replied:
will clear

aciidb0mb3r commented on Sources/Basic/FileSystem.swift:
```diff
@@ -101,6 +101,9 @@ public protocol FileSystem {
     /// Check whether the given path is accessible and a file.
     func isFile(_ path: AbsolutePath) -> Bool
+    
+    /// Check to determine whether the given path is an accessible and executable file
```
Please end all comments with punctuation.

heckj replied:
thanks, will fix

aciidb0mb3r commented on Sources/SourceControl/GitRepository.swift:
```diff
@@ -600,6 +600,17 @@ private class GitFileSystemView: FileSystem {
+            return false
+        } catch {
+            return false
+        }
```
I think this can be simplified to:
```swift
if let entry = try? getEntry(path), entry.type == .executableBlob {
    return true
}
return false
```

heckj replied:
the 3 preceding bits of logic all follow that same pattern - update them all to keep it consistent? (`isDirectory`, `isFile`, and `isSymlink`) - use the `try?` on all of them instead of the longer `do {} catch {}` block structure?

aciidb0mb3r commented on Sources/Utility/Process.swift:
```diff
@@ -373,6 +373,98 @@ extension Process {
+///   - pathString: The path string to parse.
+///   - currentWorkingDirectory: The current working directory, the relative paths will be converted to absolute paths based on this path.
+/// - Returns: List of search paths.
+public func getEnvSearchPaths(
```
Please propose this move in a separate if this PR is going to have additional commits

heckj replied:
another PR for the migration of the search paths bits from UserControls -> Process coming up

heckj replied:
moved environment search path logic (and related tests) to another PR: https://github.com/apple/swift-package-manager/pull/996

aciidb0mb3r commented on Tests/BasicTests/FileSystemTests.swift:
```diff
@@ -42,10 +43,23 @@ class FileSystemTests: XCTestCase {
+        stream <<< "set -e" <<< "\n"
+        stream <<< "exit" <<< "\n"
+        try! localFileSystem.writeFileContents(executable, bytes: stream.bytes)
+        _ = try! Process.popen(args: "chmod", "+x", executable.asString)
```
use `Process.checkNonZeroExit()` because this will not fail if chmod exits with non zero code.

heckj replied:
Ah! Thank you!

heckj commented:
@aciidb0mb3r PR updated after the various merges - still need to write tests, but wanted to get your eyes on this first. Was tough choosing where to put in a relevant check to validate `git` that wasn't in 1000 places, but also wasn't checked when not needed. I aimed for a putting it in `getActiveWorkspace` (which looked like it was always in the right flow for the places where this check would be valuable) with the idea that any of the tool's running `runImpl` would leverage that method and get the check, and I could throw an error from there, and update the Error handling output in Commands/Error.swift to print appropriately. This sound like an acceptable plan of attack, and stubbed/initial implementation?

aciidb0mb3r commented on Sources/Commands/SwiftTool.swift:
```diff
@@ -203,6 +204,17 @@ public class SwiftTool<Options: ToolOptions> {
+        }
+        guard localFileSystem.isExecutableFile(gitLocation) else {
+            throw SwiftToolError.missingGitDependency
+        }
```
We should place this check in Process class (just before calling posix_spawn) so every subprocess we spawn is checked and not make it specialised for git.

heckj replied:
I think this may be a preferrable solution, not because it's specific to git, but because it only happens once and the beginning of any invocation, regardless of how many times "git" is invoked - which in a larger project could be quite a number of times. Vapor lists 6 dependencies for example - 5 times more checks than is strictly needed to validate Git is around and in the path as an executable. 

If you insist that we should do this check as a part of Process, I'll make the changes and updates, but I'd like to ask that you reconsider.

aciidb0mb3r replied:
Thats a good point but I still think this belongs in Process class, we can cache the results so we don't need to check in subsequent calls. I'll leave it upto you if you want to implement he cache right now (it can just be a simple dictionary).

heckj replied:
Gotcha, can and will do, thanks! I'll shift it over and set up the relevant testing.

aciidb0mb3r commented on Sources/Commands/UserToolchain.swift:
```diff
@@ -57,21 +57,19 @@ public struct UserToolchain: Toolchain, ManifestResourceProvider {
-        //        (it could be a symlink to such as file).
-        guard localFileSystem.exists(swiftCompiler) else {
+        // Check that it's valid in the file system and executable.
+        guard localFileSystem.isExecutableFile(swiftCompiler) else {
```
Please propose completely unrelated changes (such as this and lookup rename) in a different PR.

heckj replied:
They didn't seem unrelated to me (the process of adding extra guards/checks to the code), but happy to move to another PR

aciidb0mb3r replied:
Ah I see, thats a fair categorization. I was thinking about this PR as a feature addition to Process class (which I still do, see previous comment).

heckj replied:
No worries, easy to shift this to another PR

heckj replied:
moved these updates to https://github.com/apple/swift-package-manager/pull/1003

heckj commented:
@aciidb0mb3r I think it's ready for a review - I can rebase to a single commit before merging, but wanted to get feedback from here. 

I leveraged the other PR's adding in the various guards, and this wraps around Process to check for errors using that simple cache mechanism. I've also added some messaging and handling into Commands/Error.swift to cover the execution failure within Process

aciidb0mb3r commented on Sources/Utility/Process.swift:
```diff
@@ -165,6 +172,24 @@ public final class Process: ObjectIdentifierProtocol {
+        guard let executableLocation = Utility.lookupExecutablePath(filename: program, searchPaths: envSearchPaths) else {
+            throw Process.Error.missingProgram(program: program)
+        }
+        guard localFileSystem.isExecutableFile(executableLocation) else {
```
I think lookupExecutable ensure this.

heckj replied:
if you mean `lookupExecutablePath`, then unfortunately no - it doesn't. It's poorly named, but I've left the name in place (perhaps I should update it?) - that method only iterates through the search paths to find the executable, it doesn't actually *verify* it's executable as well.

I could combine these together and update that function, and return a single error (missing or not executable), but with them separate I can be more specific about the error. I honestly wasn't sure which was better, so i aimed for more information to debug/understand the problem if one happens.

aciidb0mb3r replied:
In general, feel free to rename a methods to clarify their intents.

For this specific case, I actually don't want a specific error because there could be a lot of reason for failure... it could be a dir, we might not have permission, the mode of file may be invalid. Its probably better to just say "we couldn't find the executable you wanted" even tho thats a bit vague. What do you think?

heckj replied:
Based on my own past experiences of not quite understanding why something failed, I was leaning towards being as absolutely explicit about the error as I could be, presuming that for a CLI, a more detailed error would be useful to someone getting their development or CI environment debugged, and in the case of a library that the additional error types would be more useful. 

That said, the only places currently using this are in UserToolchain and now here in process, so if the specificity doesn't seem really needed, I can collapse it down to a "missing or not executable" pretty easily and shorten even this checking code. I'll update `Utility/Misc.swift` to have `lookupExecutablePath` only return a value if it's both found & executable, which can reduce the number of subsequent guards here and simplify this a bit.

aciidb0mb3r commented on Sources/Utility/Process.swift:
```diff
@@ -165,6 +172,24 @@ public final class Process: ObjectIdentifierProtocol {
+            throw Process.Error.programNotExecutable(program: program)
+        }
+        // add the program name to the validated cache
+        Process.validatedExecutables.insert(program)
```
Hmm this cache needs to be thread safe. We should also store the result, right now we will recheck for the executables we did not find before.

heckj replied:
I'm not entirely sure what you mean by "store the result", given that line 190 is appending the name of the program searched for, and found, in the Set. Do you mean store the result of where we found the executable? Even though we don't pass that back or use it? Or is that something we should be doing?

I thought about the thread-safe, and worst case that I could spot was that extra checking would happen, and the Set might be updated with an additional redundant value (and subsequently ignore it I presume). I'm feeling a little bait-and-switched with this implementation though, because my original proposal wouldn't have had to concern itself so much with thread safety since it was very early and simply in the execution path and before the system would start to attempt to do "interesting things" in parallel for efficiency. So - that said - I'm also not immediately clueful about *how* to make this thread safe, could you perhaps suggest a specific path to enabling that? Are you thinking that we should wrap the shared state (the cache) in a lock/protection akin to what is described at https://www.bignerdranch.com/blog/locking-in-swift-helping-protect-me-from-myself/ or some of the answers at http://stackoverflow.com/questions/28191079/create-thread-safe-array-in-swift?

aciidb0mb3r replied:
By store the result I meant, store the result of the lookup query. Since we don't need the actual path, we can store the result as a bool where true means we found it, false means we did not find it.

For now, lets leave the cache and it can be an add on later. I'll point you to examples in SwiftPM for thread safety.

heckj replied:
cool -I'd like that. Something I'm not yet really understanding down to how to implement it, but I'd like to.

aciidb0mb3r commented on Sources/Utility/Process.swift:
```diff
@@ -165,6 +172,24 @@ public final class Process: ObjectIdentifierProtocol {
     }
 
+    /// Validate that the program provided exists on the search path and is executable.
+    internal func validateArgumentIsExecutable(_ program: String) throws {
```
Prefer not to explicitly write internal.

heckj replied:
What would be preferred? I made it `internal` so that I could test it - when I started, I was making it `private`, but then even @testable import Utility couldn't resolve it to test it (a bit surprising to me philosophically - I'm used to testing private methods in other languages). I could just make it public, but it didn't seem like it necessarily fit in that scenario. Very open to feedback though - what do you think would make sense here?

aciidb0mb3r replied:
oh I meant that the default scope is internal and we don't explicitly write internal in swiftpm's codebase.
For e.g., these two structs have similar scope:
```swift
struct Foo {}
```
```swift
internal struct Foo{}
```

heckj replied:
/me facepalm - got it.

aciidb0mb3r commented on Sources/Utility/Process.swift:
```diff
@@ -151,6 +155,9 @@ public final class Process: ObjectIdentifierProtocol {
     /// Queue to protect concurrent reads.
     private let serialQueue = DispatchQueue(label: "org.swift.swiftpm.process")
 
+    /// cache of validated executables arguments
```
Please use proper capitalization, punctuation in all comments. http://llvm.org/docs/CodingStandards.html#commenting

heckj replied:
Ah, thank you for the link on the conventions, will update

aciidb0mb3r commented on Tests/UtilityTests/ProcessTests.swift:
```diff
@@ -95,6 +96,35 @@ class ProcessTests: XCTestCase {
         }
     }
 
+    func testProcessExecutableLookup() throws {
```
Please add this in allTests variable otherwise it won't run on linux.

heckj replied:
will do, sorry - missed that

aciidb0mb3r commented on Tests/UtilityTests/ProcessTests.swift:
```diff
@@ -95,6 +96,35 @@ class ProcessTests: XCTestCase {
+                try Process().validateArgumentIsExecutable("nonExistantProgram")
+                XCTFail("Should have failed to validate nonExistantProgram")
+            } catch Process.Error.missingProgram (let program) {
+                XCTAssertTrue(program.contains("nonExistantProgram"))
```
this can check for equality, right?

heckj replied:
duh, yes - of course it can. Will update

aciidb0mb3r commented on Sources/Utility/Process.swift:
```diff
@@ -109,6 +110,9 @@ public final class Process: ObjectIdentifierProtocol {
         /// Missing arguments to invoke a process
         case missingArguments
+        case missingProgram(program: String)
+        case programNotExecutable(program: String)
```
Please add doc comments to these

heckj replied:
will do

heckj commented:
@aciidb0mb3r simplified and updated, including the lookupExecutable method in misc and corresponding tests. Also rebased the whole kit to a single commit. I haven't make this explicitly thread-safe, so no improvements on that front from earlier commits, but I think it's in good order as it stands.

xwu commented on Sources/Utility/Process.swift:
```diff
@@ -107,8 +108,10 @@ public final class Process: ObjectIdentifierProtocol {
-        /// Missing arguments to invoke a process
+        /// The Process is missing arguments to invoke a process.
         case missingArguments
+        /// The program requested to be executed can not be found on the existing search paths, or is not executable.
```
Drive-by nits: "cannot" is one word.

xwu replied:
Drive-by nits: "cannot" is one word.

xwu commented on Sources/Utility/Process.swift:
```diff
@@ -151,6 +154,9 @@ public final class Process: ObjectIdentifierProtocol {
     /// Queue to protect concurrent reads.
     private let serialQueue = DispatchQueue(label: "org.swift.swiftpm.process")
 
+    /// Cache of validated executables arguments to be checked againstfrom `validateArgumentIsExecutable`.
```
"againstfrom"?

xwu commented on Sources/Utility/Process.swift:
```diff
@@ -165,6 +171,21 @@ public final class Process: ObjectIdentifierProtocol {
+    /// Validate that the program provided exists on the search path and is executable.
+    func validateArgumentIsExecutable(_ program: String) throws {
+
+        // check to see if the argument is already listed in the vaildated cache
```
These comments should probably be correctly capitalized and punctuated.

xwu commented on Sources/Utility/Process.swift:
```diff
@@ -165,6 +171,21 @@ public final class Process: ObjectIdentifierProtocol {
+        guard Utility.lookupExecutablePath(filename: program, searchPaths: envSearchPaths) != nil else {
+            throw Process.Error.missingExecutableProgram(program: program)
+        }
+        // add the program name to the validated cache since it has been found
```
Ditto.

xwu commented on Tests/UtilityTests/ProcessTests.swift:
```diff
@@ -95,6 +96,35 @@ class ProcessTests: XCTestCase {
+            // should always work
+            try! Process().validateArgumentIsExecutable("ls")
+
+            // should report missingProgram to non-existant program
```
Typo: "nonexistent"

xwu commented on Tests/UtilityTests/ProcessTests.swift:
```diff
@@ -95,6 +96,35 @@ class ProcessTests: XCTestCase {
+            try localFileSystem.writeFileContents(tempExecutable, bytes: "#!/bin/sh\nexit\n")
+
+            try withCustomEnv(["PATH": path.asString]) {
+                // should report nonExecutable
```
typo: "non-executable"

xwu commented on Tests/UtilityTests/miscTests.swift:
```diff
@@ -63,9 +65,10 @@ class miscTests: XCTestCase {
             XCTAssertEqual(exec, pathEnvClang)
             
-            // Create the binary relative to cwd.
+            // Create the binary relative to cwd and make it executable
```
Punctuation.

heckj commented:
thanks @xwu appreciate the call-outs - updated.

aciidb0mb3r commented on Sources/Utility/Process.swift:
```diff
@@ -165,6 +171,21 @@ public final class Process: ObjectIdentifierProtocol {
+    func validateArgumentIsExecutable(_ program: String) throws {
+
+        // Check to see if the argument is already listed in the vaildated cache.
+        guard !Process.validatedExecutables.contains(program) else {
```
Please remove the cache because it is not thread safe. It is better to do more than be unsafe,  as it can cause subtle bugs and crashes.

heckj replied:
Will do

aciidb0mb3r commented on Sources/Commands/Error.swift:
```diff
@@ -151,7 +151,10 @@ private func _handle(_ error: Any) {
-        
+
+    case Process.Error.missingExecutableProgram(let program):
+        print(error: "Unable to find an executable \(program).")
```
It's actually weird but we don't capitalise the errors (or end them with punctuation). I don't know the reason but it's consistent with compiler.

heckj replied:
Huh, OK - will update

heckj replied:
looking at the rest of this file, it's pretty mixed - there's a fair bit of initial capitalization, although much fewer punctuation ending the errors. I removed the punctuation, but left the capitalization since it seemed to fit with more other lines.

aciidb0mb3r commented on Sources/Utility/Process.swift:
```diff
@@ -165,6 +171,21 @@ public final class Process: ObjectIdentifierProtocol {
+        guard !Process.validatedExecutables.contains(program) else {
+            return
+        }
+        let envSearchPaths = Utility.getEnvSearchPaths(pathString: getenv("PATH"), currentWorkingDirectory: currentWorkingDirectory)
```
actually getenv and currentworkingdir is also not thread safe, i'll comment here again on how to make this thread safe.

heckj replied:
Cool, thanks

aciidb0mb3r replied:
I haven't tried compiling this but we need to do something similar to this:
```swift
    /// Queue to protect reading/writing on executables map.
    private static let executablesQueue = DispatchQueue(
            label: "org.swift.swiftpm.process.findExecutable")

    /// Contains results of executable lookup.
    ///
    /// Key: Executable name or path.
    /// Value: If key was found in the search paths.
    static private var executablesMap: [String: Bool] = [:]

    /// Returns true if the given program is present in search path.
    ///
    /// The program can be executable name, relative path or absolute path.
    func findExecutable(_ program: String) -> Bool {
        return Process.executablesQueue.sync {
            // Check if we already have a value for the program.
            if let value = Process.executablesMap[program] {
                return value
            }
            // FIXME: This can be cached.
            let envSearchPaths = Utility.getEnvSearchPaths(
                pathString: getenv("PATH"),
                currentWorkingDirectory: currentWorkingDirectory
            )
            // Lookup the executable.
            let value = Utility.lookupExecutablePath(
                filename: program, searchPaths: envSearchPaths) != nil
            Process.executablesMap[program] = value
            return value
        }
    }

```

heckj replied:
Thanks @aciidb0mb3r, I'll update to work out doing this - and see about adding the caching for the path lookups as well.

So the general mechanism here, if I'm groking it, is to create a DispatchQueue and set it up as a class property, so there's just one of these in the system, and then when we need to run something that leveraging things needing thread safety, to do that relevant work within a call to the DispatchQueue `sync` method, handing in an executable block that returns a value.

I started reading through the guides on DispatchQueues, and the code examples there were in C and talked to creating either Serial or Concurrent quyeyeqs. - I wasn't clear what the initializer in swift returned - does that by default create a serial one? I'll keep reading, but thought I'd ask if you any good pointers to more info here.

aciidb0mb3r replied:
The type of variable (class or property) depends on the use case, since we require that this method should be protected across all the instance of Process, it should be a class property.

Yes, by default, a serial queue is created and then you can dispatch work items on those queues. I haven't read this one specifically but tutorials here are generally good: https://www.raywenderlich.com/148513/grand-central-dispatch-tutorial-swift-3-part-1

heckj commented on Sources/Utility/Process.swift:
```diff
@@ -165,6 +178,35 @@ public final class Process: ObjectIdentifierProtocol {
+    }
+
+    /// Validate that the program provided exists on the search path and is executable.
+    func validateArgumentIsExecutable(_ program: String) throws {
```
I wasn't sure how (or if) to mix throwing exceptions with methods that are invoking synchronous commands on a DispatchQueue, I'm presuming the guard would still be valuable, but is it worth a method here to make it explicitly testable? Or would it make more sense to shift this down to a guard statement just directly in the `launch` method below, and update the tests to validate the boolean results of the search extraneous of testing for throwing exceptions to be handled? Do you have a preference for how that's structured?

aciidb0mb3r replied:
We should move the guard to launch() and just write test for `findExecutable()` and `launch()`

heckj commented:
I suspected that might be the path you suggested. I'll update

heckj commented:
@aciidb0mb3r give it another look-see? Changes made, methods collapsed, and rebased onto latest master into a single commit.

aciidb0mb3r commented:
@swift-ci please smoke test

aciidb0mb3r commented:
@swift-ci please smoke test

aciidb0mb3r commented:
@heckj Thanks for working on this!

heckj commented:
Thank you for the guidance and pointers on threadsafety!
